pkg/controllers/management/usercontrollers/controller.go

```diff
@@ -63,10 +70,239 @@ func (c *ClusterLifecycleCleanup) Create(obj *v3.Cluster) (*v3.Cluster, error) {
 }
 
 func (c *ClusterLifecycleCleanup) Remove(obj *v3.Cluster) (*v3.Cluster, error) {
+	if obj.Status.Driver == v3.ClusterDriverImported {
+		err := c.cleanupImportedCluster(obj)
+		if err != nil {
+			if apiError, ok := err.(*httperror.APIError); !ok {
+				return nil, err
+			} else {
+				if apiError.Code != httperror.ClusterUnavailable {
+					return nil, err
+				}
+			}
+		}
+	}
+
 	c.Manager.Stop(obj)
 	return nil, nil
 }
 
 func (c *ClusterLifecycleCleanup) Updated(obj *v3.Cluster) (*v3.Cluster, error) {
 	return nil, nil
 }
+
+func (c *ClusterLifecycleCleanup) cleanupImportedCluster(cluster *v3.Cluster) error {
+	userContext, err := c.Manager.UserContext(cluster.Name)
+	if err != nil {
+		return err
+	}
+
+	role, err := c.createCleanupClusterRole(userContext)
+	if err != nil {
+		return err
+	}
+
+	sa, err := c.createCleanupServiceAccount(userContext)
+	if err != nil {
+		return err
+	}
+
+	crb, err := c.createCleanupClusterRoleBinding(userContext, role.Name, sa.Name)
+	if err != nil {
+		return err
+	}
+
+	job, err := c.createCleanupJob(userContext, sa.Name)
+	if err != nil {
+		return err
+	}
+
+	or := []metav1.OwnerReference{
+		metav1.OwnerReference{
+			APIVersion: "batch/v1",
+			Kind:       "Job",
+			Name:       job.Name,
+			UID:        job.UID,
+		},
+	}
+
+	// These resouces need the ownerReference added so they get cleaned up after
+	// the job deletes itself.
+	sleepTime := 100
+	for i := 0; i <= 3; i++ {
```

Comment:
can we extract each cleanup to separate methods? Also as we seem to be using the retry logic quite often, see if it's possible to create some util method accepting function as a call back. Something similar to https://github.com/rancher/rke/blob/master/k8s/k8s.go#L68

Reply:
Yes, will add a helper for the retry and functions for each resource